HeterogeneousCore/SonicCore/interface/SonicClientTypes.h

```diff
@@ -0,0 +1,27 @@
+#ifndef HeterogeneousCore_SonicCore_SonicClientTypes
+#define HeterogeneousCore_SonicCore_SonicClientTypes
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+
+//this base class exists to limit the impact of dependent scope in derived classes
+template <typename InputT, typename OutputT = InputT>
+class SonicClientTypes {
+public:
+  //typedefs for outside accessibility
+  typedef InputT Input;
+  typedef OutputT Output;
+  //destructor
+  virtual ~SonicClientTypes() {}
+
+  //accessors
+  Input& input() { return input_; }
+  const Input& input() const { return input_; }
+  void setInput(const Input& inp) { input_ = inp; }
+  const Output& output() const { return output_; }
+
+protected:
+  Input input_;
```

Comment:
I understand that these cannot be pointers because this class acts kind of like a buffer between the possibly asynchronous back-end, and the user code that produces the input (and could go out of scope before it is sent to the backend) and consumes the output (that it may not have allocated space for yet). Is it correct ?

Reply:
Yes, that's correct. With the interface of `SonicEDProducer`, the user code just sets/gets these entries directly (passed by reference). In addition, in some cases these values might need to be accessed in other threads (PseudoAsync case).